### PR TITLE
Js restart integration

### PR DESCRIPTION
**Pull request by jsingle**
@thieman @zacharym

### CODE REVIEW

File: tests/integration/cli/restart_test.py [hunk 1]
```diff
@@ -0,0 +1,71 @@
...
+from dusty.source import Repo
+from ...testcases import DustyIntegrationTestCase
+from ...fixtures import specs_fixture_with_depends
+
+class TestRestartCLI(DustyIntegrationTestCase):
+    def setUp(self):
+        super(TestRestartCLI, self).setUp()
+        specs_fixture_with_depends()
+        for repo in get_all_repos(include_specs_repo=False):
+            self._set_up_fake_local_repo(path=repo.remote_path)
```
Comment (zacharym) [inline tests/integration/cli/restart_test.py]: might want to rename this if it's not private anymore.. i guess it is private to the class not sure what convention is on that

File: tests/integration/cli/restart_test.py [hunk 2]
```diff
@@ -0,0 +1,71 @@
...
+        self.run_command('bundles activate bundle-a bundle-b')
+        self.run_command('up')
+        time.sleep(1)
```
Comment (zacharym) [inline tests/integration/cli/restart_test.py]: why is this required?

File: tests/integration/cli/restart_test.py (hunk 1 above)
Comment (thieman) [inline tests/integration/cli/restart_test.py]: It's still private, it's only being used from within the class.

File: tests/integration/cli/restart_test.py [hunk 3]
```diff
@@ -0,0 +1,71 @@
...
+            self.run_command('stop')
+        except Exception:
+            pass
+        super(TestRestartCLI, self).tearDown()
+
+    @nottest
+    def container_has_restarted(self, app_name):
+        inspected = self.inspect_container(app_name)
+        start_time = dateutil.parser.parse(inspected['State']['StartedAt']).replace(tzinfo=None)
+        return start_time > self.up_complete_time
```
Comment (thieman) [inline tests/integration/cli/restart_test.py]: This should be an assert function defined in the test case.

File: tests/integration/cli/restart_test.py [hunk 4]
```diff
@@ -0,0 +1,71 @@
...
+
+    def test_restart_one(self):
+        self.run_command('restart appa')
+        self.assertTrue(self.container_has_restarted('appa'))
+        self.assertTrue(not self.container_has_restarted('appb'))
```
Comment (thieman) [inline tests/integration/cli/restart_test.py]: Much more idiomatic to just use `assertFalse`

File: tests/integration/cli/restart_test.py (hunk 2 above)
Comment (zacharym) [inline tests/integration/cli/restart_test.py]: oh for `has_restarted`.. hmm

Comment (thieman) [inline tests/integration/cli/restart_test.py]: Doesn't `StartedAt` give us much more than second precision?

File: tests/integration/cli/restart_test.py [hunk 5]
```diff
@@ -0,0 +1,71 @@
...
+
+    def test_restart_by_app_repo(self):
+        self.run_command('restart --repos app-a')
+        self.assertTrue(self.container_has_restarted('appa'))
+        self.assertTrue(not self.container_has_restarted('appb'))
+
+    def test_restart_by_lib_repo(self):
+        self.run_command('restart --repos lib-a')
+        self.assertTrue(self.container_has_restarted('appa'))
+        self.assertTrue(self.container_has_restarted('appb'))
```
Comment (zacharym) [inline tests/integration/cli/restart_test.py]: Might be worth it to add an app that doesn't depend on this lib and confirm that it hasn't restarted

Comment (thieman) [inline tests/integration/cli/restart_test.py]: +1

File: tests/fixtures.py [hunk 6]
```diff
@@ -49,6 +49,44 @@ def basic_specs_fixture():
                             'mount': '/lib/a',})
     _write('service', 'service-a', {'image': 'service/a'})
 
+def specs_fixture_with_depends():
+    _write('bundle', 'bundle-a', {'description': 'Bundle A', 'apps': ['appa']})
+    _write('bundle', 'bundle-b', {'description': 'Bundle B', 'apps': ['appb']})
+    _write('app', 'appa', {'repo': '/tmp/app-a',
```
Comment (thieman) [inline tests/fixtures.py]: Can we call the repo `repo-a` or something? Would make the tests much more clear

File: tests/testcases.py [hunk 7]
```diff
@@ -136,11 +136,12 @@ def run_command(self, args, fake_exit, fake_exec_docker):
         self._clear_stdout()
         return result
 
-    def _set_up_fake_local_repo(self):
-        repo = git.Repo.init('/tmp/fake-repo')
-        with open('/tmp/fake-repo/README.md', 'w') as f:
+    def _set_up_fake_local_repo(self, path='/tmp/fake-repo'):
+        print "PATH: {}".format(path)
```
Comment (zacharym) [inline tests/testcases.py]: debugging?

File: tests/integration/cli/restart_test.py [hunk 8]
```diff
@@ -0,0 +1,71 @@
...
+
+    def test_restart_nosync(self):
+        new_file_name = 'nosync_file'
+        repo = Repo.resolve(get_all_repos(include_specs_repo=False), 'app-a')
+        with open(os.path.join(repo.local_path, new_file_name), 'w+') as f:
+            f.write('new file!')
+        self.run_command('restart --no-sync appa')
+        self.assertFileNotInContainer('appa', os.path.join('/app/a/', new_file_name))
+
+    def test_restart_nosync(self):
```
Comment (thieman) [inline tests/integration/cli/restart_test.py]: Naming drep

File: tests/testcases.py [hunk 9]
```diff
@@ -188,6 +189,14 @@ def assertFileContents(self, file_path, contents):
     def assertFileContentsInContainer(self, service_name, file_path, contents):
         self.assertEqual(self.exec_in_container(service_name, 'cat {}'.format(file_path)), contents)
 
+    def assertFileNotInContainer(self, service_name, file_path):
+        self.assertEqual(self.exec_in_container(service_name,
+            'sh -c \'[ -f {} ] && echo "yes" || echo "no"\''.format(file_path)).rstrip(), 'no')
+
+    def assertFileInContainer(self, service_name, file_path):
+        self.assertEqual(self.exec_in_container(service_name,
+            'sh -c \'[ -f {} ] && echo "yes" || echo "no"\''.format(file_path)).rstrip(), 'yes')
```
Comment (thieman) [inline tests/testcases.py]: These should share code between each other, see some of the other examples in the class.

File: tests/testcases.py [hunk 10]
```diff
@@ -208,3 +217,9 @@ def assertImageDoesNotExist(self, image):
 
     def assertExecDocker(self, *args):
         self.fake_exec_docker.assert_called_with(*args)
+
+    def inspect_container(self, service_name):
+        client = get_docker_client()
+        container_id = _get_container_for_app_or_service(client, service_name, include_exited=True)['Id']
+        return client.inspect_container(container_id)
```
Comment (thieman) [inline tests/testcases.py]: Can you put this definition above the asserts for the sake of keeping some sort of structure? It's a big hairy class already.

**Comment by thieman** [comment]
Design's good, bunch of nitpicks.

File: tests/integration/cli/restart_test.py (hunk 2 above)
Comment (jsingle) [inline tests/integration/cli/restart_test.py]: It gives many milisecond digits, but on at least one occasion this method was unreliable without the sleep.  Perhaps `run_command('up')` can return before the start time per container is set?  I could look into that possibility

File: tests/integration/cli/restart_test.py (hunk 3 above)
Comment (jsingle) [inline tests/integration/cli/restart_test.py]: Not sure what you mean.  The test cases assert that this function returns true or false

File: tests/integration/cli/restart_test.py (hunk 2 above)
Comment (thieman) [inline tests/integration/cli/restart_test.py]: Can we do a shorter sleep?

File: tests/testcases.py (hunk 7 above)
Comment (jsingle) [inline tests/testcases.py]: Yep

**Comment by jsingle** [comment]
Makin changes

File: tests/integration/cli/restart_test.py (hunk 2 above)
Comment (jsingle) [inline tests/integration/cli/restart_test.py]: Very likely

File: tests/integration/cli/restart_test.py (hunk 3 above)
Comment (jsingle) [inline tests/integration/cli/restart_test.py]: It would be nice to leave the functions in this file, so that the setup code can stay in this subclass as well

Comment (thieman) [inline tests/integration/cli/restart_test.py]: wat

Comment (jsingle) [inline tests/integration/cli/restart_test.py]: Talking about setting `self.up_complete_time` and the `time.sleep`
